topology/probes/ovs_of.go

```diff
@@ -0,0 +1,531 @@
+	if err != nil {
+		close(cout)
+		return cout, err
+	}
+	reader := bufio.NewReader(out)
+	var line string
+
+	go func() {
+		for {
+			line, err = reader.ReadString('\n')
+			if err == io.EOF {
+				close(cout)
+				break
+			} else if err != nil {
+				logging.GetLogger().Errorf("IO Error on command %s: %s", cmd[0], err.Error())
+			} else {
+				cout <- line
+			}
+		}
+	}()
+
+	return cout, nil
+}
+
+func countElements(filter string) int {
+	if len(filter) == 0 {
+		return 1
+	}
+	elts := strings.Split(filter, ",")
+	l := len(elts) + 1
+	for _, elt := range elts {
+		if strings.HasPrefix(elt, "priority=") {
+			l = l - 1
+			break
+		}
+	}
+	return l
+}
+
+// completeRule completes the rule by looking at it again but with dump-flows. This gives back more elements such as priority.
+func completeRule(o *OvsOfProbe, event *Event) error {
+	oldrule := event.Rule
+	bridge := event.Bridge
+	// We want exactly n+1 items where n was the number of items in old filters. The reason is that now
+	// the priority is provided. Another approach would be to use the shortest filter as it is the more generic
+	expected := countElements(oldrule.Filter)
+	filter := makeFilter(oldrule)
+	command, err1 := o.makeCommand([]string{"ovs-ofctl", "dump-flows"}, bridge, filter)
+	if err1 != nil {
+		return err1
+	}
+	lines, err := LaunchOnSwitch(command)
+	if err != nil {
+		return fmt.Errorf("Cannot launch ovs-ofctl dump-flows on %s@%s with filter %s: %s", bridge, o.Host, filter, err.Error())
+	}
+	done := false
+	for _, line := range strings.Split(lines, "\n") {
+		rule, err2 := parseRule(line)
+		if err2 == nil && countElements(rule.Filter) == expected && oldrule.Cookie == rule.Cookie {
+			if done {
+				logging.GetLogger().Errorf("Multiple completion for rule on %s@%s with filter %s", bridge, o.Host, filter)
+			}
+			oldrule.Filter = rule.Filter
+			oldrule.Actions = rule.Actions
+			done = true
+		}
+	}
+	extractPriority(oldrule)
+	if !done {
+		return fmt.Errorf("Cannot complete rule on %s@%s with filter %s", bridge, o.Host, filter)
+	}
+	return nil
+}
+
+// addRule adds a rule to the graph and links it to the bridge.
+func addRule(rule *Rule, probe *BridgeOfProbe) {
+	g := probe.OvsOfProbe.Graph
+	g.Lock()
+	defer g.Unlock()
+	bridgeNode := probe.BridgeNode
+	name := rule.Filter + " => " + rule.Actions
+	metadata := graph.Metadata{
+		"Type":     "ofrule",
+		"Name":     name,
+		"cookie":   fmt.Sprintf("0x%x", rule.Cookie),
+		"table":    rule.Table,
+		"filters":  rule.Filter,
+		"actions":  rule.Actions,
+		"priority": rule.Priority,
+		"UUID":     rule.UUID,
+	}
+	ruleNode := g.NewNode(graph.GenID(), metadata)
+	g.Link(bridgeNode, ruleNode, graph.Metadata{"RelationType": "ownership"})
+}
+
+// delRule deletes a rule from the the graph.
+func delRule(rule *Rule, probe *BridgeOfProbe) {
+	g := probe.OvsOfProbe.Graph
+	g.Lock()
+	defer g.Unlock()
+
+	ruleNode := g.LookupFirstNode(graph.Metadata{"UUID": rule.UUID})
```

Comment:
I am not listeneing to del-flows but to what ovs-ofctl monitor watch: gives back. And it will give back the exact filter of all the rules removed each on a separate line.